### ISSUE TITLE
Custom Attribute (**Implementation available**)

### ISSUE THREAD

**Issue by WhileTrueEndWhile** [reinert/JJSchema#87]
Hello,

through a targeted search for "POJO to JSON schema" I found JJSchema.
I would like to generate JSON schemas from Java classes, which are then used to create forms.
Unfortunately, non-standard metadata must be added for this.
With `@Attributes` I don't see any way around the given attribute names.

I imagined something like this:

```
// ...
    @Attributes(

        title = "Hello World",

        custom = new String[][] {
            // new String[] { "KEY", "VALUE" }
            new String[] { "widget", "select" }
            // ...
        }

    )
// ...
```

Whereby `widget` should better be called `x-widget`, but you can't always decide that yourself...
Specifically I would like to use it together with `ngx-schema-form` :-)

If the annotation is attached to an attribute of a class, the JSON scheme might look something like this:

```
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "type": "object",
    "properties": {
        "...":  {
            "type": "...",
            // ...
            // "KEY": "VALUE" <---- new String[] { "KEY", "VALUE" }
            "widget": "select" // <---- new String[] { "widget", "select" }
        }
        // ...
    }
    // ...
}
```

I don't expect a complete solution, but I just don't know if it is possible without a fork.

Thanks in advance!

**Comment by reinert** [comment]
Hi there. There's no such feature. Feel free to implement it and propose a patch.

**Comment by WhileTrueEndWhile** [comment]
Thank you for the quick answer.

As far as I can see, `@Attributes` is the data model to modify and `JsonSchemaGenerator` is the implementation to modify. Using the following call sequence for fields, you probably have to adapt `processSchemaProperty`:

1. `checkAndProcessType`
2. `processCustomType`
3. `processProperties`
4. `generatePropertySchema`
5. `processSchemaProperty`

The method `processSchemaProperty` is abstract and is implemented in `JsonSchemaGeneratorV4`.
It looks like only one method of implementation needs to be changed.

I would want to adjust the in `JsonSchemaGenveratorV4`, but this is `@Deprecated`.
Is this still the right place?

**Comment by WhileTrueEndWhile** [comment]
Since the development will certainly take some time, I would like to introduce my idea of the API...

# JSON Schema Extension API for JJSchema

Provides an interface to maintain user-defined properties in a JSON schema generated by JJSchema or to overwrite or remove already set properties, if necessary.

```
String[][] x() default null
```

The annotation `@Attribute` is extended by `x`. where `x` is a list of lists.
The inner list is a data object, where the last value is the actual value to be set and the other values describe the path.

```
Path =  ArrayItem_0, ArrayItem_1, ArrayItem_2, ... ArrayItem_(m-1)
Value = ArrayItem_m
```

All values of the inner list are parsed. Thus, other primitive values are also assigned by predetermined string values.

```
@Attributes(x = new String[][]{
    //                |---------------- Path -----------------|  |-- Value --|
    new String[][] {
        new String[]{ "Key_0", "Key_1", "Key_2", "...", "Key_n", "Hello World" }
        // ...
    }
)
```

- This means that the value `Hello World` remains the same.
- If a key or value takes the value `42`, it is converted into an `Integer`.
- If the value is `true` or `false`, it becomes a `Boolean`.
- A value of `null`  deletes the entry in the schema.

## Example I: Simple

```
@Attributes(x = new String[][]{
    //                String  String  String       /            JSON Schema
    //                  |       |       |         /     vvvvvvvvvvvvvvvvvvvvvvvvvvvv
    // new String[]{ "widget", "id", "number" } // => { "widget": { "id": "number" } }
})
```

## Example II: Complex

```
new String[][] {
    //               String   Integer  String   String        /  JSON Schema
    //                  |        |        |        |         /   vvvvvvvvvvv
    // new String[]{ "Key_0", "Key_1", "Key_2", "Value" }   /
    //                                                     /   { ..., "fieldsets": [
    new String[]{ "fieldsets", "0", "field", "Foo" }    // =>     { "field": "Foo" },
    new String[]{ "fieldsets", "1", "field", "Bar" }    // =>     { "field": "Bar" }
    // ...                                               //    ], ...}
}
```

# TODOs

- [ ] Setup development environment ;-)
- [ ] Add `String[][] x() default null` to `@Attribute`
- [ ] Implement string parsing for keys and values (string to other primitive)
- [ ] Implement insertion strategy (string keys create objects, number keys create arrays)
- [ ] Apply code and write test
- [ ] Create patch

**Comment by reinert** [comment]
> Is this still the right place?

Use [JsonSchemaV4Factory](https://github.com/reinert/JJSchema/blob/master/src/main/java/com/github/reinert/jjschema/v1/JsonSchemaV4Factory.java) instead.

> `String[][] x() default null`

Name it `x-properties` or `x-props`, please.

**Comment by WhileTrueEndWhile** [comment]
Ok, I revised the documentation.

**Comment by WhileTrueEndWhile** [comment]
Hi again,

I have created a first patch. I had to adjust the documentation again, because `String[][]` is not allowed for annotations, but `String[]` is allowed:

https://gist.githubusercontent.com/WhileTrueEndWhile/e3bbbf4e0d9b7d77771de0ee8dc04842/raw/9ad9d21e713a309a5ba3b913427ccc30b31dacd9/xproperties.patch

Well, it looks pretty good, but I don't know how to integrate it into the V1, because I want to add a central method (`processSchemaProperty`), which probably doesn't exist anymore...
I need the `@Attributes` instance and the current `JsonNode`!

Then I can write a few tests^^

Thanks!
